ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,605 @@
+  * Workflow can customize its name via `workflow_name <name>` in the workflow somewhere.
+  * **NOTE** As part of this proposal, we are changing the activity base class from `Temporalio::Activity` class to
+    `Temporalio::Activity::Definition` to be consistent with the base class for workflows.
+  * 💭 Why a class?
+    * This fits well with the Ruby mindset and many of our SDKs utilize classes to encapsulate workflows and their
+      handlers/state.
+* Workflow has several class methods that start with `workflow_` that can only be called during class definition time
+  not runtime. See next section for list.
+* When used in a client:
+  * The class of the workflow can be provided instead of the name, e.g. `my_client.start_workflow(MyWorkflowClass, ...)`
+  * The signal, query, or update can reference a no-arg class method of the same name as the instance method, e.g.
+    `my_wf_handle.signal(MyWorkflowClass.my_signal, ...)`, which returns a signal/query/update definition
+    * Are we ok with `respond_to_missing?` and `inherited` for class method access of these values? Is this too much
+      magic?
+* Workflow cannot be instantiated by users (i.e. outside of workflow environment)
+
+## Workflow API
+
+Approximate guess of how the `Workflow` module may look:
+
+```ruby
+module Temporalio
+  module Workflow
+
+    # Class users must extend for a workflow
+    class Definition
+      class << self
+        # Protected because they are only for this implementation
+        protected
+
+        # Placed above handlers. See dynamic section for dynamic. Raw args means the
+        # handlers will be given their arguments as RawValue instead of converted.
+        #
+        # ❓ Notice raw_args. We could instead make type hints this way, e.g. via
+        # `args: [String, Integer]` or something and plumb that all the way through
+        # to the converter for a user's custom converter to use (same for activities
+        # and the workflow as a whole). If we did that we would also need to allow
+        # return type hints. Do we want to support this hacky form of type hints?
+        def self.workflow_signal(name: nil, dynamic: false, raw_args: false,
+                                unfinished_policy: HandlerUnfinishedPolicy::WARN_AND_ABANDON)
+        def self.workflow_query(name: nil, dynamic: false, raw_args: false)
+        def self.workflow_update(name: nil, dynamic: false, raw_args: false,
+                                unfinished_policy: HandlerUnfinishedPolicy::WARN_AND_ABANDON)
+        def self.workflow_update_validator(update_method)
+
+        # Exposes an attribute as a query. 💭 Why? We found this _very_ valuable in
+        # .NET. Think of it as attr_reader for workflows. It was also decided there
+        # was not enough value to have attr_accessor shortcut for update or
+        # attr_writer shortcut for signals since both are not commonly just
+        # getter/setter.
+        def self.workflow_query_attr(query_method)
```

Comment:
This is cool.
nit: I wonder if `workflow_query_attr_reader` would more grokkable by the ruby community, despite the partial reduncancy of query vs reader.

Reply:
:+1: Can add suffix

Reply:
Added